extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs

```diff
@@ -80,47 +81,42 @@ private CosmosDBInputAttribute GetBindingDataContent(ModelBindingData bindingDat
-            _ => await CreateTargetObjectAsync(targetType, cosmosAttribute)
+            _ => ConvertDynamic(await CreateTargetObjectAsync(targetType, cosmosAttribute), targetType)
         };
 
-        private async Task<object> CreateTargetObjectAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute)
+        private object ConvertDynamic(dynamic data, Type targetType)
         {
-            MethodInfo createPOCOMethod;
-
-            if (targetType.GenericTypeArguments.Any())
-            {
-                targetType = targetType.GenericTypeArguments.FirstOrDefault();
-
-                createPOCOMethod = GetType()
-                                    .GetMethod(nameof(CreatePOCOCollectionAsync), BindingFlags.Instance | BindingFlags.NonPublic)
-                                    .MakeGenericMethod(targetType);
-            }
-            else
-            {
-                createPOCOMethod = GetType()
-                                    .GetMethod(nameof(CreatePOCOAsync), BindingFlags.Instance | BindingFlags.NonPublic)
-                                    .MakeGenericMethod(targetType);
-            }
-
+            var json = JsonSerializer.Serialize(data);
+            return JsonSerializer.Deserialize(json, targetType, _serializerOptions);
+        }
 
+        private async Task<object> CreateTargetObjectAsync(Type targetType, CosmosDBInputAttribute cosmosAttribute)
+        {
             var container = CreateCosmosClient<Container>(cosmosAttribute) as Container;
 
             if (container is null)
             {
                 throw new InvalidOperationException($"Unable to create Cosmos container client for '{cosmosAttribute.ContainerName}'.");
             }
 
-            return await (Task<object>)createPOCOMethod.Invoke(this, new object[] { container, cosmosAttribute });
+            if (targetType.GenericTypeArguments.Any())
+            {
+                return await CreatePOCOCollectionAsync(container, cosmosAttribute);
+            }
+            else
+            {
+                return await CreatePOCOAsync(container, cosmosAttribute);
+            }
         }
 
-        private async Task<object> CreatePOCOAsync<T>(Container container, CosmosDBInputAttribute cosmosAttribute)
+        private async Task<object> CreatePOCOAsync(Container container, CosmosDBInputAttribute cosmosAttribute)
         {
             if (String.IsNullOrEmpty(cosmosAttribute.Id) || String.IsNullOrEmpty(cosmosAttribute.PartitionKey))
             {
                 throw new InvalidOperationException("The 'Id' and 'PartitionKey' properties of a CosmosDB single-item input binding cannot be null or empty.");
             }
 
-            ItemResponse<T> item = await container.ReadItemAsync<T>(cosmosAttribute.Id, new PartitionKey(cosmosAttribute.PartitionKey));
+            ItemResponse<dynamic> item = await container.ReadItemAsync<dynamic>(cosmosAttribute.Id, new PartitionKey(cosmosAttribute.PartitionKey));
```

Comment:
I think we can use `ReadItemStreamAsync`, pass back the `Stream` from `ResponseMessage.Content` and deserialize that via `JsonSerializer`. This should be more performant than using `dynamic`.

Reply:
I didn't even think to check for a Stream method - nice :D